Intro/README.md

```diff
@@ -0,0 +1,455 @@
+of untrusted code will not be reflected in any RTM.
+
+## Object Naming
+
+TPMs support a variety of types of objects.  Objects generally have
+pointer-like "handles" that they are often used in the TPM APIs.  But
+more importantly, objects have cryptographically-secure names that are
+used in many cases.
+
+  The cryptographically-secure name of an object is the hash of the
+  object's "public area".
+
+The public area of, say, an asymmetric key is a data structure that
+includes the public key (corresponding to the private key), and various
+attributes of the key (e.g., its algorithm, but also whether it is bound
+to the TPM where it resides, or to its key hierarchy), unseal policy,
+and access policy.
+
+This concept is extremely important.  Because object names are the
+outputs of cryptographically strong digest (hash) functions, they are
+resistant to collision attacks, first pre-image attacks, and second
+pre-image attacks -- as strong as the hash algorithm used anyways.
+Which means that object names cannot be forged easily, which means that
+they can be used in context where a peer needs certain guarantees, or
+to defeat active attacks.
+
+### Immutability of Object Public Areas
+
+Because the name of an object is a digest of its public area, the public
+area cannot be changed after creating it.  One can delete and then
+recreate an object in order to "change" its public area, but this
+necessarily yields a new name (assuming no digest collisions).
+
+## Key Hierarchies
+
+TPMs have multiple key hierarchies, all rooted in a primary decrypt-only
+asymmetric private key derived from a seed, with arbitrarily complex
+trees of keys below the primary key:
+
+```
+                seed
+                 |
+                 |
+                 v
+     primary key (asymmetric encryption)
+                 |
+                 |
+                 v
+       secondary keys (of any kind)
+                 |
+                 |
+                 v
+                ...
+```
+
+There are three built-in hierarchies:
+
+ - platform hierarchy
+ - endorsement hierarchy
+ - storage hierarchy
+
+of which only the endorsement and storage hierarchies will be of
+interest to most readers.
+
+The endorsement hierarchy is used to authenticate (when needed) that a
+TPM is a legitimate TPM.  The primary endorsement key is known as the EK
+(endorsement key).  Hardware TPMs come with a certificate for the EK
+issued by the TPM's manufacturer.  This EK certificate ("EKcert") can be
+used to authenticate the TPM's legitimacy.  The EK's public key
+("EKpub") can be used to uniquely identify a TPM, and possibly link to
+the platform's, and even the platform's user(s)' identities.
+
+## Key Wrapping and Resource Management
+
+The primary key is always a decrypt-only asymmetric private key, and its
+corresponding public key is therefore encrypt-only.  This is largely
+because of key wrapping, where a symmetric key or asymmetric private key
```

Comment:
Maybe instead of "symmetric key or asymmetric private key" just say "private key" here? It took me a few tries to parse this sentence.

Reply:
Hmm, to me "private key" implies asymmetric crypto, and "secret key" implies symmetric crypto.  I'll re-phrase.

Reply:
Inaccurate.  A primary key can be whatever you want it to be.  It doesn't even have to be a key.

Reply:
@kgold2 hmm, but in the EK hierarchy the primary key has to be a decrypt-only key, no?

Reply:
> 
> 
> @kgold2 hmm, but in the EK hierarchy the primary key has to be a decrypt-only key, no?

There is no EK hierarchy.
There is an endorsement hierarchy, and I can create any primary object with any attributes I like.

I encourage you to download the IBM TSS.  You can test these assertions in seconds.